tests/backup/backup_restore_basic_test.go

```diff
@@ -2229,3 +2230,224 @@ var _ = Describe("{MultipleInPlaceRestoreSameTime}", func() {
+				credName = fmt.Sprintf("autogenerated-cred-%v", time.Now().Unix())
+				credMap[provider] = make(map[string]string)
+				credMap[provider][cloudCredUID] = credName
+				err := CreateCloudCredential(provider, credName, cloudCredUID, orgID, ctx)
+				dash.VerifyFatal(err, nil, fmt.Sprintf("Verifying creation of cloud credential named [%s] for org [%s] with [%s] as provider", credName, orgID, provider))
+				log.InfoD("Created Cloud Credentials with name - %s", credName)
+				customBackupLocationName = fmt.Sprintf("autogenerated-backup-location-%v", time.Now().Unix())
+				err = CreateBackupLocation(provider, customBackupLocationName, backupLocationUID, credName, cloudCredUID, getGlobalBucketName(provider), orgID, "")
+				dash.VerifyFatal(err, nil, fmt.Sprintf("Creating backup location %s", customBackupLocationName))
+				backupLocationMap[backupLocationUID] = customBackupLocationName
+			}
+		})
+
+		Step("Register source and destination cluster for backup", func() {
+			log.InfoD("Registering Source and Destination clusters and verifying the status")
+			ctx, err := backup.GetAdminCtxFromSecret()
+			log.FailOnError(err, "Fetching px-central-admin ctx")
+			err = CreateSourceAndDestClusters(orgID, "", "", ctx)
+			dash.VerifyFatal(err, nil, "Creating source and destination cluster")
+			clusterStatus, err = Inst().Backup.GetClusterStatus(orgID, SourceClusterName, ctx)
+			log.FailOnError(err, fmt.Sprintf("Fetching [%s] cluster status", SourceClusterName))
+			dash.VerifyFatal(clusterStatus, api.ClusterInfo_StatusInfo_Online, fmt.Sprintf("Verifying if [%s] cluster is online", SourceClusterName))
+			clusterUid, err = Inst().Backup.GetClusterUID(ctx, orgID, SourceClusterName)
+			dash.VerifyFatal(err, nil, fmt.Sprintf("Fetching [%s] cluster uid", SourceClusterName))
+		})
+
+		Step("Taking backup of applications", func() {
+			log.InfoD("Taking backup of applications")
+			var sem = make(chan struct{}, numberOfSimultaneousBackups)
+			var wg sync.WaitGroup
+			ctx, err := backup.GetAdminCtxFromSecret()
+			log.FailOnError(err, "Fetching px-central-admin ctx")
+			log.InfoD("Taking %d backups", numberOfBackups)
+			for backupLocationUID, backupLocationName := range backupLocationMap {
+				for _, namespace := range bkpNamespaces {
+					for i := 0; i < numberOfBackups; i++ {
+						sem <- struct{}{}
+						time.Sleep(timeBetweenConsecutiveBackups)
+						backupName := fmt.Sprintf("%s-%v", BackupNamePrefix, time.Now().Unix())
+						backupNames = append(backupNames, backupName)
+						wg.Add(1)
+						go func(backupName string) {
+							defer GinkgoRecover()
+							defer wg.Done()
+							defer func() { <-sem }()
+							err = CreateBackup(backupName, SourceClusterName, backupLocationName, backupLocationUID, []string{namespace},
+								labelSelectors, orgID, clusterUid, "", "", "", "", ctx)
+							dash.VerifyFatal(err, nil, fmt.Sprintf("Verifying backup creation: %s", backupName))
+						}(backupName)
+					}
+				}
+				wg.Wait()
+			}
```

Comment:
@tthurlapati-px If I go by what you are saying why is it not failing for all cases https://github.com/portworx/torpedo/blob/master/tests/backup/backup_restore_basic_test.go#L1212 ? The same code is being using at multiple places and seem to be working fine without whatever changes you are suggesting.

Reply:
In the particular example you have shown over here the **flaw does exist**, but it doesn't manifest because `wg.wait()` is being used such that `namespace` variable will remain the same until all corresponding goroutines are done, thus averting a race condition.

But the example code can also be further optimized and made faster by moving `wg.Wait()` outside the outer for loop, **and** make the change to goroutine as `go func(backupName, namespace string) {` in order to prevent race condition.

So, my comments on the current code are still valid

Reply:
@tthurlapati-px I get your point but there is no point of fixing it in just a single place. We can't do all the optimisation in a single PR. We can take this forward in the next one and do it at all places. I see you have approved this PR 3 times and went back and opened new comments 3 times. This is of no use and the PR will always be in an infinite loop forever and never get merged. We can check if whatever you are saying works and then we can do these changes at all places instead of doing it at just once place which doesn't really help.

Reply:
While I understand what you're saying about the _optimization_ part of things, even if you don't optmize the `wg.Wait()`, there **still** exists a bug that you may have slipped past you. 

If you compare the example code you gave and your own current code, you'll notice that the example code ensures that `wg.Wait()` prevents race conditions as `namespace` doesn't change until all goroutine are complete.

But in your code, `namespace` **does** change before all goroutines complete, keeping the window open for race conditions. I'm not sure why you haven't hit an error. (is it because of limiting the numberofsimultneousbackups?)

Reply:
I approved the PR.
The optimizations can be done later 👍🏻